src/spatial/nsm.jl

```diff
@@ -0,0 +1,269 @@
+end
+
+#NOTE integrator and params are not used. They remain to adhere to the interface of `AbstractSSAJumpAggregator` defined in ssajump.jl
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::NSMJumpAggregation, integrator, u, params, t)
+    fill_rates_and_get_times!(p, u, t)
+    generate_jumps!(p, integrator, params, u, t)
+    nothing
+end
+
+#NOTE integrator and params are not used. They remain to adhere to the interface of `AbstractSSAJumpAggregator` defined in ssajump.jl
+# calculate the next jump / jump time
+function generate_jumps!(p::NSMJumpAggregation, integrator, params, u, t)
+    @unpack cur_rates, rng = p
+
+    p.next_jump_time, site = top_with_handle(p.pq)
+    if rand(rng)*get_site_rate(cur_rates, site) < get_site_reactions_rate(cur_rates, site)
+        rx = linear_search(get_site_reactions_iterator(cur_rates, site), rand(rng) * get_site_reactions_rate(cur_rates, site))
+        p.next_jump = SpatialJump(site, rx+length(@view p.diffusion_constants[:,site]), site)
+    else
+        species_to_diffuse = linear_search(get_site_diffusions_iterator(cur_rates, site), rand(rng) * get_site_diffusions_rate(cur_rates, site))
+        #TODO this is not efficient. We iterate over neighbors twice.
+        n = rand(rng,1:num_neighbors(p.spatial_system, site))
+        target_site = nth_neighbor(p.spatial_system,site,n)
+        p.next_jump = SpatialJump(site, species_to_diffuse, target_site)
+    end
+end
+
+# execute one jump, changing the system state
+function execute_jumps!(p::NSMJumpAggregation, integrator, u, params, t)
+    # execute jump
+    update_state!(p, integrator)
+
+    # update current jump rates and times
+    update_dependent_rates_and_firing_times!(p, integrator.u, t)
+    nothing
+end
+
+######################## SSA specific helper routines ########################
+"""
+reevaluate all rates, recalculate tentative site firing times, and reinit the priority queue
+"""
+function fill_rates_and_get_times!(aggregation::NSMJumpAggregation, u, t)
+    @unpack ma_jumps, diffusion_constants, spatial_system = aggregation
+
+    num_majumps = get_num_majumps(ma_jumps)
+    num_species = length(@view u[:,1]) #NOTE assumes u is a matrix with ith column being the ith site
+    num_sites = number_of_sites(spatial_system)
+    cur_rates = SpatialRates(num_majumps,num_species,num_sites)
+
+    @assert cur_rates.reaction_rates_sum == zeros(typeof(cur_rates.reaction_rates_sum[1]),num_sites)
+    @assert cur_rates.diffusion_rates_sum == zeros(typeof(cur_rates.diffusion_rates_sum[1]),num_sites)
+
+    pqdata = Vector{typeof(t)}(undef, num_sites)
+    for site in 1:num_sites
+        update_reaction_rates!(cur_rates, 1:num_majumps, u, ma_jumps, site)
+        update_diffusion_rates!(cur_rates, 1:num_species, diffusion_constants, u, site, spatial_system)
+        pqdata[site] = t + randexp(aggregation.rng) / get_site_rate(cur_rates, site)
+    end
+
+    aggregation.cur_rates = cur_rates
```

Comment:
Again, don’t allocate just use the existing allocate field.

Reply:
okay